### ISSUE TITLE
RTKQ: Optimistic rendering. Doubt and discussion

### ISSUE THREAD

**Issue by lukascivil** [reduxjs/redux-toolkit#1246]
I would like to know if there is any way for "optimistic rendering" or if it makes sense to be something implemented by the lib. Who would be responsible for the strategy, who calls (the App) or the lib itself?

I'll leave an example here, but it's not complete, I think it can cause problems with the types and not make much sense, but the idea is to illustrate how it could be.

#### Example
ListComponent.tsx
```tsx
 const { data, refetch, isFetching } = useGetUsersQuery(
    {
      filter: {},
      pagination: { page: 1, perPage: 5 },
      sort: currentSort
    },
    { refetchOnMountOrArgChange: 2, pollingInterval: 10000 }
  )

// Response
[{ id: 1, ...}, {id:2, ...}, ...]
```

DetailsComponent.tsx (Detail of any component of the listing)
```tsx
// It is important to afirm here that I know the element id but not the parameters used to find it.
  const { data: user, refetch, initialValue } = useGetUserQuery(2, {
    refetchOnMountOrArgChange: true,
  // It will be used for optimistic rendering, and will return anything from the state. Must be run before the request, and return my initial value. 
 // The idea here is to bring element 2 that was already loaded into memory by another call.
 // Another option, selectFromStateFirst: (state) => state.data.user[id] || "..."
   selectFromStateFirst: selectUser(2) 
  })

// Response
{id:2, ...}
```

####  How am i doing today
DetailsComponent.tsx
```tsx
  const { data: cachedUser } = useGetLoadedUserQuery({ id }, { refetchOnMountOrArgChange: true })
  const { data: user, refetch } = useGetUserQuery(id, {
    refetchOnMountOrArgChange: true
  })

// Response
{id:2, ...}
```

users-api.ts
```tsx
export const usersApi = createApi({
...
   getLoadedUser: build.query<User, { id: number; target?: string }>({
      queryFn: ({ id, target }, api) => {
        const state = api.getState()

        // cachedItem will be used for optimistic rendering
        const cachedItem = selectRtkCachedItem<User>(state, usersApi.reducerPath, { id, target })

        return cachedItem
      }
    })
})
```

Thanks for the effort in RTK query, the tool is amazing.

**Comment by phryneas** [comment]
Can you please elaborate what you mean by "optimistic rerendering"?

I honestly have never heard that term before.

**Comment by lukascivil** [comment]
@phryneas Ok. This is not "optimistic re-rendering", the idea is to take a user (example) that was loaded into memory by a listing method (getUsers()) on another screen.

Imagine you are on a screen that lists 100 users and you want to view the details of a specific user. You know the user is already loaded into memory, why not take advantage of that? And, make another specific call, which will bring the user again, because it may have been updated.

Basically this strategy will avoid a loading spinner. So, the problem is based more or less on the previous case, if you have any tips, thanks.

**Comment by phryneas** [comment]
For that, you would need to have the argument you sent to `useGetUsersQuery` somewhere in the state
```js
const arg = useSelector(state => state.foo.bar.useGetUsersQueryArg)
const { data: userFromCache } = api.endpoints.getUsers.useQueryState(arg, { 
  selectFromResult: result => ({ data: result.data.users.find(user => user.id == id) })
})
const { data: user, refetch } = useGetUserQuery(id, {
  refetchOnMountOrArgChange: true,
  skip: userFromCache !== undefined
})
```

**Comment by shurshilov** [comment]
In fact, optimistic rendering is a popular thing. It visually allows you to speed up the interface quite a lot. This is used where the data rarely changes, we can immediately draw the form if we previously requested data from the list. Or if we store data in the local storage, then we also have the effect of automatic rendering when data is restored from it and an instant blurring effect is created, and then loading occurs in the background. For example, this is implemented in react admin) Here is my simple code example for a form in react native that works great.
`  // optimistic render
  const { data: dataCache } = odooApi.endpoints.searchRead.useQueryState(
    {
      model: 'res.partner',
      context: {},
      domain: [],
      fields: ['name', 'id', 'function', 'phone'],
      limit: 1,
    },
    {
      selectFromResult: result => ({
        data: result.data?.find(user => user.id === route.params.id),
      }),
    },
  );`
